src/Kestrel.Core/Internal/Http2/Http2Connection.cs

```diff
@@ -672,7 +672,15 @@ private void StartStream<TContext>(IHttpApplication<TContext> application)
             }
 
             _streams[_incomingFrame.StreamId] = _currentHeadersStream;
-            _ = _currentHeadersStream.ProcessRequestsAsync(application);
+            // Must not allow app code to block the connection handling loop.
+            System.Threading.ThreadPool.UnsafeQueueUserWorkItem(state =>
+                {
+                    var tuple = (Tuple<IHttpApplication<TContext>, Http2Stream>)state;
+                    var app = tuple.Item1;
+                    var currentStream = tuple.Item2;
+                    _ = currentStream.ProcessRequestsAsync(app);
+                },
+                new Tuple<IHttpApplication<TContext>,Http2Stream>(application, _currentHeadersStream));
```

Comment:
Nit: For more @pranavkm points you can shorten this to

```C#
ThreadPool.UnsafeQueueUserWorkItem(state =>
{
    var (app, currentStream) = (Tuple<IHttpApplication<TContext>, Http2Stream>)state;
    _ = currentStream.ProcessRequestsAsync(app);
},
(application, _currentHeadersStream));
```

Reply:
With `UnsafeQueueUserWorkItem` the state-argument (valuetuple) gets boxed. This allocation can be avoided with the generic `QueueUserWorkItem`-overload:
```c#
ThreadPool.QueueUserWorkItem(state =>
{
    var (app, currentStream) = state;
    _ = currentStream.ProcessRequestsAsync(app);
},
(application, _currentHeadersStream), false);
```
Or is the "unsafe queuing" needed here?

Further the allocation for the delegate could be avoided by caching it.

Reply:
Unsafe is more efficient as it avoids capturing the current thread context.

Reply:
I'm not able to fully extract the delegate due to the method's generic type.

Reply:
> Unsafe is more efficient as it avoids capturing the current thread context.

Can always do this instead for a similar effect
```csharp
bool restoreFlow = false;
try
{
    if (!ExecutionContext.IsFlowSuppressed())
    {
        ExecutionContext.SuppressFlow();
        restoreFlow = true;
    }

    ThreadPool.QueueUserWorkItem(state =>
    {
        var (app, currentStream) = state;
            _ = currentStream.ProcessRequestsAsync(app);
    },
    (application, _currentHeadersStream), false);
}
finally
{
    // Restore the current ExecutionContext
    if (restoreFlow)
        ExecutionContext.RestoreFlow();
}
```
As per https://github.com/dotnet/corefx/pull/26077

Reply:
What's the advantage to doing that over UnsafeQueueUserWorkItem?

Reply:
Depends if you can use the generic state overload (on netcoreapp2.1) to avoid the boxing of the state
```csharp
QueueUserWorkItem<TState>(Action<TState> callBack, TState state, bool preferLocal)
```

Reply:
Also is an `Action<T>` callback rather than a `WaitCallback`

Reply:
We're trying to limit our use of netcoreapp2.1 APIs to only places where we can't avoid it. If this shows up in the profiles then we may try adding a fast-path.

Reply:
Is no advantage vs `UnsafeQueueUserWorkItem` on netstandard2.0

Reply:
Besides "limit our use of netcoreapp2.1 APIs" would it make sense to make `UnsafeQueueUserWorkItem` generic similar to https://github.com/dotnet/coreclr/pull/16570?

Reply:
I was suggesting the new syntax for terseness not for valuetuple. We don't want to ifdef to use netcoreapp2.1 APIs unless theres a functional improvement or significant perf improvement. Explicitly call ExecutionContext.Suppress/RestoreFlow is also pretty gross.